### ISSUE TITLE
Different behaviour for optional decoders depending on whether they're at top level

### ISSUE THREAD

**Issue by hgiddens** [circe/circe#851]
Given a `Decoder[A]`, trying to decode an `Option[A]` will fail differently (assuming the underlying `Decoder[A]` fails) depending on whether we're looking at the top level or not:

```
import io.circe.{Decoder, Json}

Decoder.
  instance(_.as[Option[Int]]).
  decodeJson(Json.fromString(""))
// Right(None)

Decoder.
  instance(_.downField("foo").as[Option[Int]]).
  decodeJson(Json.obj(("foo", Json.fromString(""))))
// Left(DecodingFailure(Int, List(DownField(foo))))
```

Is this behaviour deliberate? And if so is there an easy way out of it (so that the first example would also fail with the `Left`)?

Tested with 0.9.1 and current master; the behaviour from a quick glance looks like it's always been this way though.

**Comment by travisbrown** [comment]
I'd categorize this as more "known" than "deliberate". It's behavior that's inherited from Argonaut, although Argonaut 6.2 no longer does this.

I think I agree that the only thing that should decode into a `None: Option[A]` at the top-level is a JSON `null`, and would be happy to see this change in 0.10. PR welcome, otherwise I'm happy to handle it this weekend.

**Comment by hgiddens** [comment]
I'll have a go at fixing it up.

**Comment by travisbrown** [comment]
Fixed in #853.
